File: affine.js
function cifradoAfin(texto, a, b) {
    let cifrado = "";
    for (let i = 0; i < texto.length; i++) {
        let char = texto[i];
        if (char.match(/[a-z]/i)) {
            let codigo = char.charCodeAt(0);
            if (codigo >= 97 && codigo <= 122) {
                cifrado += String.fromCharCode(((a * (codigo - 97) + b) % 26) + 97);
            } else if (codigo >= 65 && codigo <= 90) {
                cifrado += String.fromCharCode(((a * (codigo - 65) + b) % 26) + 65);
            }
        } else {
            cifrado += char;
        }
    }
    return cifrado;
}

function inversoMod(a, m) {
    for (let x = 1; x < m; x++) {
        if ((a * x) % m === 1) {
            return x;
        }
    }
    return 1;
}

function descifradoAfin(texto, a, b) {
    let inversoModA = inversoMod(a, 26);
    let descifrado = "";
    for (let i = 0; i < texto.length; i++) {
        let char = texto[i];
        if (char.match(/[a-z]/i)) {
            let codigo = char.charCodeAt(0);
            if (codigo >= 97 && codigo <= 122) {
                descifrado += String.fromCharCode(((inversoModA * (codigo - 97 - b + 26)) % 26) + 97);
            } else if (codigo >= 65 && codigo <= 90) {
                descifrado += String.fromCharCode(((inversoModA * (codigo - 65 - b + 26)) % 26) + 65);
            }
        } else {
            descifrado += char;
        }
    }
    return descifrado;
}

// Ejemplo de uso
let texto = "Hola, Mundo!";
let a = 5;
let b = 8;
let cifrado = cifradoAfin(texto, a, b);
let descifrado = descifradoAfin(cifrado, a, b);
console.log("Cifrado:", cifrado);
console.log("Descifrado:", descifrado);

File: affine.py
def cifrado_afin(texto, a, b):
    cifrado = ""
    for char in texto:
        if char.isalpha():
            if char.islower():
                cifrado += chr((a * (ord(char) - ord('a')) + b) % 26 + ord('a'))
            elif char.isupper():
                cifrado += chr((a * (ord(char) - ord('A')) + b) % 26 + ord('A'))
        else:
            cifrado += char
    return cifrado

def descifrado_afin(texto, a, b):
    inverso_mod_a = pow(a, -1, 26)
    descifrado = ""
    for char in texto:
        if char.isalpha():
            if char.islower():
                descifrado += chr((inverso_mod_a * (ord(char) - ord('a') - b)) % 26 + ord('a'))
            elif char.isupper():
                descifrado += chr((inverso_mod_a * (ord(char) - ord('A') - b)) % 26 + ord('A'))
        else:
            descifrado += char
    return descifrado

# Ejemplo de uso
texto = "Hola, Mundo!"
a = 5
b = 8
cifrado = cifrado_afin(texto, a, b)
descifrado = descifrado_afin(cifrado, a, b)
print("Cifrado:", cifrado)
print("Descifrado:", descifrado)
